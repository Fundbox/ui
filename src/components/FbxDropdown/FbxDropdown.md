# FbxDropdown
A dropdown menu with a slot the content of the dropdown itself.

Intentionally, the dropdown itself is almost unstyled (besides for positioning) 
to make this component as customizable as possible. 
So you have to take care if styling the inside of the dropdown yourself.
Basically this component includes the down arrow button that opens up the slot 
content in an absolutely positioned container.  

### Usage
```html
<template>
  <div class="drop-down-wrapper">
    Some Menu text
    <fbx-dropdown>
      <a href="#">Help & Support</a>
      <a href="#">Agreement</a>
      <a href="#">Logout</a>
    </fbx-dropdown>
  </div>
</template>

<script>
import FbxDropdown from '../FbxDropdown.vue'
export default {
  components: { FbxDropdown },
}
</script>

<style lang="scss" scoped>
@import "./../../../styles/utils/color-palette.scss";
a {
  padding: 3px 12px;
  color: $medium-blue;
  white-space: nowrap;
  font-weight: 300;
  font-size: 14px;
  display: block;

  &:last-child {
    border-top: 1px solid $medium-gray;
  }

  &:hover {
    background: $medium-blue;
    color: white;
  }
}
</style>

<style lang="scss">
@import "./../../../styles/utils/color-palette.scss";
.drop-down-content {
  border: 1px solid $medium-gray;
}
</style>
```
