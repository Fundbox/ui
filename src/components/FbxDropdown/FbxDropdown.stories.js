import { storiesOf } from '@storybook/vue'
import { doc } from 'storybook-readme'

import DropDownMenu from './stories/DropDownMenu.story.vue'
import summary from './FbxDropdown.md'
import DropDownMenuWithText from './stories/DropDownMenuWithText.story'

const stories = storiesOf('Components/Dropdown', module)

stories.add('description', doc(summary))
stories.add('drop down menu', () => DropDownMenu)
stories.add('drop down menu with text', () => DropDownMenuWithText)
