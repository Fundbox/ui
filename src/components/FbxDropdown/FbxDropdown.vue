<template>
  <div class="drop-down">
    <button class="drop-down-button fbx-icon-arrow-down" @click="dropDownButtonClicked" @focusout="clickedOutside"></button>
    <transition name="slide-down">
      <div class="drop-down-content" v-if="isMenuOpen">
        <slot></slot>
      </div>
    </transition>
  </div>
</template>

<script>
export default {
  name: 'FbxDropdown',
  data() {
    return {
      isMenuOpen: false
    }
  },
  methods: {
    dropDownButtonClicked() {
      this.isMenuOpen = !this.isMenuOpen
    },
    clickedOutside() {
      // Delay closing by 200ms for UX purposes.
      setTimeout(() => this.isMenuOpen = false, 200)
    },
  }
}
</script>

<style lang="scss" scoped>
@import "../../styles/fonts/icomoon/style.css";

.drop-down {
  display: inline-block;
  position: relative;
}

.drop-down-button {
  cursor: pointer;
}

.drop-down-content {
  $positionFromTop: 120%;
  position: absolute;
  top: $positionFromTop;
  background: white;
  opacity: 1;
  transition: all 200ms ease-in-out;

  &.slide-down-enter, &.slide-down-leave-to {
    opacity: 0;
    top: 100%;
  }

  &.slide-down-enter-to, &.slide-down-leave {
    opacity: 1;
    top: $positionFromTop;
  }
}
</style>
