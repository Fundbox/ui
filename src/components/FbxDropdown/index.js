import FbxDropdown from './FbxDropdown.vue'
import { registerComponents, vueUse } from '../../utils/plugins'

const components = {
  FbxDropdown
}

const VuePlugin = {
  install(Vue) {
    registerComponents(Vue, components)
  }
}

vueUse(VuePlugin)

export default VuePlugin
