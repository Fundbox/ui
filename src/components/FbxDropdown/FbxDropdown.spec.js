import { shallowMount } from '@vue/test-utils'
import FbxDropdown from './FbxDropdown.vue'

describe('Components/FbxDropdown', () => {
  describe('snapshots', () => {
    it('renders the default correctly', () => {
      const wrapper = shallowMount(FbxDropdown, {})
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders menu with text', () => {
      const wrapper = shallowMount(FbxDropdown, {
        props: {
          menuText: 'menu text',
        },
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders the open menu', () => {
      const wrapper = shallowMount(FbxDropdown, {})
      wrapper.find('.drop-down-button').trigger('click')
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('menu closed when clicked outside',(done) => {
      const wrapper = shallowMount(FbxDropdown, {})
      wrapper.find('.drop-down-button').trigger('click')
      wrapper.find('.drop-down-button').trigger('focusout')
      setTimeout(() => {
        expect(wrapper.html()).toMatchSnapshot()
        done()
      }, 500)
    })
  })
})
