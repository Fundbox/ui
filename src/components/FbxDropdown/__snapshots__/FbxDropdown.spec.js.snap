// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/FbxDropdown snapshots menu closed when clicked outside 1`] = `
<div class="drop-down"><button class="drop-down-button">

    <!----> <span class="menu-icon fbx-icon-arrow-down"></span></button>
  <!---->
</div>
`;

exports[`Components/FbxDropdown snapshots renders menu with text 1`] = `
<div class="drop-down"><button class="drop-down-button">

    <!----> <span class="menu-icon fbx-icon-arrow-down"></span></button>
  <!---->
</div>
`;

exports[`Components/FbxDropdown snapshots renders the default correctly 1`] = `
<div class="drop-down"><button class="drop-down-button">

    <!----> <span class="menu-icon fbx-icon-arrow-down"></span></button>
  <!---->
</div>
`;

exports[`Components/FbxDropdown snapshots renders the open menu 1`] = `
<div class="drop-down"><button class="drop-down-button">

    <!----> <span class="menu-icon fbx-icon-arrow-down"></span></button>
  <div class="drop-down-content" name="slide-down"></div>
</div>
`;
